common/malloc_utils/src/jemalloc.rs

```diff
@@ -124,6 +125,22 @@ pub fn page_size() -> Result<usize, Error> {
     "arenas.page\0".name().read()
 }
 
+/// Uses `mallctl` to call `"prof.dump"`.
+///
+/// This generates a heap profile at `filename`.
+#[allow(dead_code)]
+pub fn prof_dump(filename: &str) -> Result<(), String> {
+    // Add a C-style `0x00` terminator to the filename and convert it to a
+    // `CStr`.
+    let mut terminated_filename = vec![0x00_u8; filename.len() + 1];
+    terminated_filename[..filename.len()].copy_from_slice(filename.as_ref());
+    let cstr = CStr::from_bytes_with_nul(&terminated_filename)
+        .map_err(|e| format!("Cannot convert filename to CStr: {e:?}"))?;
+
+    unsafe { raw::write("prof.dump".as_ref(), cstr) }
```

Comment:
LGTM, looking at the source this corresponds to `mallctl("prof.dump", NULL, NULL, filename, size)` which is exactly what we want

Ref:

- https://github.com/jemalloc/jemalloc/wiki/Use-Case%3A-Heap-Profiling
- https://docs.rs/tikv-jemalloc-ctl/0.6.0/src/tikv_jemalloc_ctl/raw.rs.html#137-147